dockerfiles/theia/src/main.js

```diff
@@ -11,46 +11,159 @@
+
+/** Returns standard (default) Theia dependencies */
+function getDefaultTheiaPlugins() {
+    const defaultTheiaPackageJson = require(`${defaultTheiaRoot}/package.json`);
+    return defaultTheiaPackageJson['dependencies'];
+}
+
+/** Returns current (after previous launch) Theia plugins or empty object on first start */
+function getTheiaPlugins() {
+    if (!fs.existsSync(`${theiaRoot}/package.json`)) {
+        return {};
+    }
+    const theiaPackageJson = require(`${theiaRoot}/package.json`);
+    return theiaPackageJson['dependencies'];
+}
+
+/** Returns extra Theia plugins which is set via THEIA_PLUGINS env variable */
+function getExtraTheiaPlugins() {
+    let plugins = {};
+    const currentPluginsString = process.env.THEIA_PLUGINS;
+    if (currentPluginsString) {
+        for (let plugin of currentPluginsString.split(',')) {
+            plugin = plugin.trim();
+            if (plugin === '') {
                 continue;
             }
-            if (!dep.hasOwnProperty(d)) {
-                dep[d] = "latest";
+            const colonPos = plugin.indexOf(':');
+            if (colonPos !== -1) {
+                const pluginName = plugin.substring(0, colonPos).trim();
+                const pluginVersion = plugin.substring(colonPos + 1).trim();
+                if (pluginVersion.startsWith('git://') || pluginVersion.startsWith('https://')) {
+                    addPluginFromGitRepository(plugins, pluginName, pluginVersion);
+                } else {
+                    plugins[pluginName] = pluginVersion;
+                }
+            } else {
+                plugins[plugin] = 'latest';
             }
         }
-        fs.writeFileSync(theiaPath, JSON.stringify(theiaConfig));
-        handlePromise(callYarn().then(callBuild).then(callRun));
-    } else {
-        cp.execSync(`rsync -rv ${defaultTheiaPath}/ ${theiaRoot} --exclude 'node_modules' --exclude 'yarn.lock'`);
-        handlePromise(callRun());
+    }
+    return plugins;
+}
+
+/**
+ * Clones git repository and adds the plugin as local.
+ * This is done becouse Theia requires npm package.json (not lerna) which is located, in most cases, in subdirectory of repository.
+ */
+function addPluginFromGitRepository(plugins, pluginName, gitRepository) {
+    const pluginPath = gitPluginsRoot + '/' + pluginName + '/';
+    try {
+        cp.execSync(`git clone --depth=1 --quiet ${gitRepository} ${pluginPath}`);
+    } catch (error) {
+        // failed to clone repository
+        try {
+            cp.execSync(`git clone --depth=1 --quiet ${gitRepository} ${pluginPath}`);
+        } catch (error) {
+            // failed again, skip plugin
+            console.error('Failed to get plugin: ' + pluginName + '. Skipping... Is the url specified properly?');
+            return;
+        }
     }
 
+    const rootPackageJson = require(pluginPath + 'package.json');
+    if (rootPackageJson['name'] === pluginName && !fs.existsSync(pluginPath + 'lerna.json')) {
```

Comment:
question: do we need to use all synchronous API or we could use async ? and then install in parallel ?

Reply:
@benoitf what we can do in parallel that's cloning of git repositories and building them if needed.
But this makes the logic of the script more complicated. We have to deal with paralleling and barrier then, which is not supported in nodejs natively. So, some kind of emulation is needed. We'll end up with batch of promises wrapped in timeouts and async await chain up to the 'main' function. More, not all the plugins are git-hosted, but to have similar code we'll have to use promises for other plugins where they aren't needed actually. But the main problem is stability of that code. Having parallelism emulation and all that stuff is error prone. Also it could be not such efficient as it looks like with this approach.
Bringing real parallelism with child process library also isn't trivial. Threads may interact via IPC or some libs for shared memory, which is require new dependency and again much more complicated and error prone code.
And taking into account, that we are going to have new plugins model - this script will be removed then. So, I think, that it is not needed for now.